PWGHF/D2H/Tasks/taskDs.cxx

```diff
@@ -545,191 +663,393 @@ struct HfTaskDs {
+              nCandsInSignalRegionDsPerType[DataType::McDplusPrompt] += checkCandInSignalRegion(candidate, true) ? 1 : 0;
+              nCandsInSignalRegionDplusPerType[DataType::McDplusPrompt] += checkCandInSignalRegion(candidate, false) ? 1 : 0;
+            } else if (isDplusNonPrompt(candidate)) {
+              ++nCandsPerType[DataType::McDplusNonPrompt];
+              nCandsInSignalRegionDsPerType[DataType::McDplusNonPrompt] += checkCandInSignalRegion(candidate, true) ? 1 : 0;
+              nCandsInSignalRegionDplusPerType[DataType::McDplusNonPrompt] += checkCandInSignalRegion(candidate, false) ? 1 : 0;
+            } else {
+              ++nCandsPerType[DataType::McBkg];
+              nCandsInSignalRegionDsPerType[DataType::McBkg] += checkCandInSignalRegion(candidate, true) ? 1 : 0;
+              nCandsInSignalRegionDplusPerType[DataType::McBkg] += checkCandInSignalRegion(candidate, false) ? 1 : 0;
+            }
+          }
+        } else {
+          auto groupedDsCandidates = candidates.sliceBy(candDsMcRecoPerCollision, thisCollId);
+          for (const auto& candidate : groupedDsCandidates) {
+            if (candidate.isSelDsToKKPi() < selectionFlagDs && candidate.isSelDsToPiKK() < selectionFlagDs)
+              continue;
+
+            if (isDsPrompt(candidate)) {
+              ++nCandsPerType[DataType::McDsPrompt];
+              nCandsInSignalRegionDsPerType[DataType::McDsPrompt] += checkCandInSignalRegion(candidate, true) ? 1 : 0;
+              nCandsInSignalRegionDplusPerType[DataType::McDsPrompt] += checkCandInSignalRegion(candidate, false) ? 1 : 0;
+            } else if (isDsNonPrompt(candidate)) {
+              ++nCandsPerType[DataType::McDsNonPrompt];
+              nCandsInSignalRegionDsPerType[DataType::McDsNonPrompt] += checkCandInSignalRegion(candidate, true) ? 1 : 0;
+              nCandsInSignalRegionDplusPerType[DataType::McDsNonPrompt] += checkCandInSignalRegion(candidate, false) ? 1 : 0;
+            } else if (isDplusPrompt(candidate)) {
+              ++nCandsPerType[DataType::McDplusPrompt];
+              nCandsInSignalRegionDsPerType[DataType::McDplusPrompt] += checkCandInSignalRegion(candidate, true) ? 1 : 0;
+              nCandsInSignalRegionDplusPerType[DataType::McDplusPrompt] += checkCandInSignalRegion(candidate, false) ? 1 : 0;
+            } else if (isDplusNonPrompt(candidate)) {
+              ++nCandsPerType[DataType::McDplusNonPrompt];
+              nCandsInSignalRegionDsPerType[DataType::McDplusNonPrompt] += checkCandInSignalRegion(candidate, true) ? 1 : 0;
+              nCandsInSignalRegionDplusPerType[DataType::McDplusNonPrompt] += checkCandInSignalRegion(candidate, false) ? 1 : 0;
+            } else {
+              ++nCandsPerType[DataType::McBkg];
+              nCandsInSignalRegionDsPerType[DataType::McBkg] += checkCandInSignalRegion(candidate, true) ? 1 : 0;
+              nCandsInSignalRegionDplusPerType[DataType::McBkg] += checkCandInSignalRegion(candidate, false) ? 1 : 0;
```

Comment:
to avoid this code duplication, you could call the `fillNPvContribHisto` function for the single collision, passing the sliced candidates and do the loop externally (but up to you if you want to modify it now or later on). 
In this case, I would also modify the `runDataAnalysis` and `runMcAnalysis` functions similarly (currently in the process functions you are looping multiple times over the collisions while you could do it only once)